AnyStr parameter + `sys.version_info` check in function causes mypy to not realize True is bool

**Bug Report**

Consider the following code:

```python
from __future__ import annotations
import sys
from typing import AnyStr


def foo(s: AnyStr) -> bool:
    if sys.version_info >= (4, 42):  # Something that won't be matched on any extant Python
        return True
    return False
```

Running `mypy --warn-return-any` on this code produces the following erroneous error:

```
bool04.py:8: error: Returning Any from function declared to return "bool"  [no-any-return]
Found 1 error in 1 file (checked 1 source file)
```

If the type of the `s` parameter is changed to `str`, or the parameter is removed entirely, then the error goes away.  The error also goes away if the `sys.version_info` check is something that succeeds on the current Python.

**Expected Behavior**

`mypy` should not have reported any errors.

**Your Environment**

- Mypy version used: 1.13.0
- Mypy command-line flags: `--warn-return-any`
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: observed on Python 3.8.20, Python 3.9.20, Python 3.12.7, and Python 3.13.0

--- a/mypy/nodes.py
+++ b/mypy/nodes.py
@@ -1263,7 +1263,7 @@ class Block(Statement):
 
     __match_args__ = ("body", "is_unreachable")
 
-    def __init__(self, body: list[Statement]) -> None:
+    def __init__(self, body: list[Statement], *, is_unreachable: bool = False) -> None:
         super().__init__()
         self.body = body
         # True if we can determine that this block is not executed during semantic
@@ -1271,7 +1271,7 @@ class Block(Statement):
         # something like "if PY3:" when using Python 2. However, some code is
         # only considered unreachable during type checking and this is not true
         # in those cases.
-        self.is_unreachable = False
+        self.is_unreachable = is_unreachable
 
     def accept(self, visitor: StatementVisitor[T]) -> T:
         return visitor.visit_block(self)

--- a/mypy/treetransform.py
+++ b/mypy/treetransform.py
@@ -276,7 +276,9 @@ class TransformVisitor(NodeVisitor[Node]):
         return NonlocalDecl(node.names.copy())
 
     def visit_block(self, node: Block) -> Block:
-        return Block(self.statements(node.body))
+        new = Block(self.statements(node.body))
+        new.is_unreachable = node.is_unreachable
+        return new
 
     def visit_decorator(self, node: Decorator) -> Decorator:
         # Note that a Decorator must be transformed to a Decorator.

--- a/mypy/treetransform.py
+++ b/mypy/treetransform.py
@@ -276,9 +276,7 @@ class TransformVisitor(NodeVisitor[Node]):
         return NonlocalDecl(node.names.copy())
 
     def visit_block(self, node: Block) -> Block:
-        new = Block(self.statements(node.body))
-        new.is_unreachable = node.is_unreachable
-        return new
+        return Block(self.statements(node.body), is_unreachable=node.is_unreachable)
 
     def visit_decorator(self, node: Decorator) -> Decorator:
         # Note that a Decorator must be transformed to a Decorator.

--- a/test-data/unit/check-unreachable-code.test
+++ b/test-data/unit/check-unreachable-code.test
@@ -1000,6 +1000,16 @@ class Test4(Generic[T3]):
 
 [builtins fixtures/isinstancelist.pyi]
 
+[case testUnreachableBlockStaysUnreachableWithTypeVarConstraints]
+# flags: --always-false COMPILE_TIME_FALSE
+from typing import TypeVar
+COMPILE_TIME_FALSE = False
+T = TypeVar("T", int, str)
+def foo(x: T) -> T:
+    if COMPILE_TIME_FALSE:
+        return "bad"
+    return x
+
 [case testUnreachableFlagContextManagersNoSuppress]
 # flags: --warn-unreachable
 from contextlib import contextmanager

